runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp

```diff
@@ -8,6 +8,127 @@
 
 namespace builtins {
 
+JSString *PenaltyBox::get_name(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+
+  return JS::GetReservedSlot(self, Slots::Name).toString();
+}
+
+// add(entry: string, timeToLive: number): void;
+bool PenaltyBox::add(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The PenaltyBox builtin");
+  METHOD_HEADER(2);
+
+  // Convert entry parameter into a string
+  auto entry = core::encode(cx, args.get(0));
+  if (!entry) {
+    return false;
+  }
+
+  // Convert timeToLive parameter into a number
+  double timeToLive;
+  if (!JS::ToNumber(cx, args.get(1), &timeToLive)) {
+    return false;
+  }
+
+  // This needs to happen on the happy-path as these all end up being valid uint32_t values that the
+  // host-call accepts
+  if (std::isnan(timeToLive) || std::isinf(timeToLive) || timeToLive < 1 || timeToLive > 60) {
+    JS_ReportErrorASCII(cx, "add: timeToLive parameter is an invalid value, only numbers from 1 to "
+                            "60 can be used for timeToLive values.");
+    return false;
+  }
+
+  // We expose it in minutes as the value gets truncated to minutes however the host expects it in
+  // seconds
+  timeToLive = timeToLive * 60;
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::PenaltyBox::add(name, entry, timeToLive);
+  if (auto *err = res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  args.rval().setUndefined();
+  return true;
+}
+
+// has(entry: string): boolean;
+bool PenaltyBox::has(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The PenaltyBox builtin");
+  METHOD_HEADER(1);
+
+  // Convert entry parameter into a string
+  auto entry = core::encode(cx, args.get(0));
+  if (!entry) {
+    return false;
+  }
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::PenaltyBox::has(name, entry);
+  if (auto *err = res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  args.rval().setBoolean(res.unwrap());
+  return true;
+}
+
+const JSFunctionSpec PenaltyBox::static_methods[] = {
+    JS_FS_END,
+};
+
+const JSPropertySpec PenaltyBox::static_properties[] = {
+    JS_PS_END,
+};
+
+const JSFunctionSpec PenaltyBox::methods[] = {JS_FN("add", add, 2, JSPROP_ENUMERATE),
+                                              JS_FN("has", has, 1, JSPROP_ENUMERATE), JS_FS_END};
```

Comment:
Ending the last entry with a comma will get `clang-format` to lay this array out a bit better.
```suggestion
                                              JS_FN("has", has, 1, JSPROP_ENUMERATE), JS_FS_END,};
```